near-sdk/src/store/free_list/mod.rs

```diff
@@ -220,6 +220,84 @@ where
     pub fn drain(&mut self) -> Drain<T> {
         Drain::new(self)
     }
+
+    /// Empty slots in the front of the list is swapped with occupied slots in back of the list.
+    /// Defrag helps reduce gas cost in certain scenarios where lot of elements in front of the list are
+    /// removed without getting replaced. Please see https://github.com/near/near-sdk-rs/issues/990
+    pub fn defrag(&mut self) -> HashMap<u32, u32> {
```

Comment:
since this is just used internally, I'd suggest an API that just does one defrag at a time, which would make it easier to use in other collections, or have this function accept a callback, that gives back the old and new indices.

If this method is kept, can just keep this in a `Vec<(u32, u32)>` to avoid the HashMap overhead here

Reply:
Replacing one element at a time might not be a good approach. If we defragmented 5 elements then while defragmenting 6th element we'll have to traverse past the first five. Also, we'll not know if the list is already defragmented or not until we traverse all free indexes.

Reply:
yeah, you're right. That was an inaccurate suggestion

Reply:
I like the idea of callback function. Would it be okay to call the callback function every swap rather than at the end?

Reply:
yes, I think that would be most sound, given that there aren't multiple swaps of the same element as discussed on the other thread.